src/accumulator/proof.rs

```diff
@@ -241,6 +243,286 @@ impl Proof {
+    fn update_proof_remove(
+        self,
+        block_targets: Vec<u64>,
+        cached_hashes: Vec<sha256::Hash>,
+        updated: Vec<(u64, sha256::Hash)>,
+        num_leaves: u64,
+    ) -> Result<(Proof, Vec<sha256::Hash>), String> {
+        let total_rows = util::tree_rows(num_leaves);
+
+        let targets_with_hash: Vec<(u64, bitcoin_hashes::sha256::Hash)> = self
+            .targets
+            .clone()
+            .into_iter()
+            .zip(cached_hashes.clone().into_iter())
+            .filter(|(pos, _)| !block_targets.contains(pos))
+            .collect();
+        let (targets, target_hashes) = targets_with_hash.to_owned().into_iter().unzip();
+
+        let proof_positions =
+            util::get_proof_positions(&self.targets, num_leaves, util::tree_rows(num_leaves));
+
+        let old_proof: Vec<_> = proof_positions
+            .iter()
+            .zip(self.hashes.clone().into_iter())
+            .collect();
+
+        let mut new_proof = vec![];
+        // Grab all the positions of the needed proof hashes.
+        let needed_pos = util::get_proof_positions(&targets, num_leaves, total_rows);
+
+        let updated: Vec<_> = updated
+            .into_iter()
+            .filter(|(pos, _)| proof_positions.contains(pos))
+            .collect();
+
+        let old_proof_iter = old_proof.into_iter();
+
+        // Loop through old_proofs and only add the needed proof hashes.
+        for (pos, hash) in old_proof_iter {
+            if let Some((_, updated_hash)) =
+                updated.iter().find(|(updated_pos, _)| pos == updated_pos)
+            {
+                if *updated_hash != sha256::Hash::default() {
+                    new_proof.push((*pos, *updated_hash));
+                }
+            } else {
+                if hash != sha256::Hash::default() {
+                    new_proof.push((*pos, hash));
+                }
+            }
+        }
+
+        let missing_positions = needed_pos
+            .into_iter()
+            .filter(|pos| proof_positions.contains(pos))
+            .peekable();
+
+        for missing in missing_positions {
+            if let Some((_, hash)) = updated
+                .iter()
+                .find(|(updated_pos, _)| missing == *updated_pos)
+            {
+                if *hash != sha256::Hash::default() {
+                    new_proof.push((missing, *hash));
+                }
+            }
+        }
+
+        // We need to remap all proof hashes and sort then, otherwise our hash will be wrong.
+        // This happens because deletion moves nodes upwards, some of this nodes may be a proof
+        // element. If so we move it to its new position. After this, we might have duplicates, and
+        // the vector is probably unsorted, so we dedup and sort it again.
+        let (pos, hashes): (Vec<u64>, Vec<sha256::Hash>) =
+            Proof::calc_next_positions(&block_targets, &new_proof, num_leaves, false)?;
```

Comment:
There shouldn't be any duplicates resulting from `calc_next_positions`.

I found that the `missing_positions` variable is wrong, leading to duplicates being added. No need to dedupe here. Maybe need to sort? I'm not sure if the code here guarantees sorted-ness.